### ISSUE TITLE
Optional custom search key

### ISSUE THREAD

**Issue by amirakbulut** [pacocoursey/cmdk#181]
First of all, thank you for all the great work.

In a lot of cases you query your API by `id` instead of `name`, because of that we set the `value` property of selectable items to be equal to their `id`.

The component however, compares search queries against this  `value` property, which is totally fine and expected, but it requires us to search our items by`id` instead of readable names. Which is not that user friendly.

Is there a way to search by a specific property, but have the `value` property still be equal to `id`?

Maybe something like `value="id" searchValue="name"`

**Comment by AntonioErdeljac** [comment]
I ran into this problem, while I have not found the props to do exactly what you describe, I used "value" a a combination of my label and unique ID.

```
<CommandDialog open={isOpen} onOpenChange={onClose}>
    <CommandInput placeholder="Type a command or search..." />
    <CommandList>
      <CommandEmpty>No results found.</CommandEmpty>
      <CommandGroup heading="Documents">
        {documents?.map((document) => (
          <CommandItem
            value={`${document._id}-${document.title}`}
            title={document.title}
            onSelect={() => onSelect(document._id)} 
            key={document._id}
          >
            <FileText className="mr-2 h-4 w-4" />
            <span>{document.title}</span>
          </CommandItem>
        ))}
      </CommandGroup>
    </CommandList>
  </CommandDialog>
```

This is definitely a step into better UX, but you can also search by "id" so users might get confused why they are seeing certain results if they type uuid for example haha

**Comment by EduartePaiva** [comment]
I'm with the same issue, I wanted to be able to search by the name, but I wanted to be able to assign the ID of the item when I the use select the item, but doing that searching is disabled. I was doing something like this,

it's in chadcn ui, but the underline code is using cmdk

`<Command>
                    <CommandInput placeholder="Procurar exercício..." />
                    <CommandEmpty>Nenhum exercício encontrado.</CommandEmpty>
                    <CommandGroup>
                        {exercicios.map((exercicio) => (
                            <CommandItem
                                key={exercicio.id}
                                onSelect={(currentValue) => {
                                    setValue(currentValue === value ? "" : currentValue)
                                    setOpen(false)
                                }}
                                value={exercicio.id}
                                title={exercicio.nome}
                            >
                                <Check
                                    className={cn(
                                        "mr-2 h-4 w-4",
                                        value === exercicio.id ? "opacity-100" : "opacity-0"
                                    )}
                                />
                                {exercicio.nome}
                            </CommandItem>
                        ))}
                    </CommandGroup>
                </Command>`

**Comment by taro-ishihara** [comment]
I think you can disable filter totally and make state based custom filter by yourself.
`<Command shouldFilter={false}>`
This is so much easier

**Comment by guizmo** [comment]
here is my solution, using Shadcd/ui

Key points"
- CommandInput : add  value={search} and onValueChange={setSearch} to empty search afterwards (on popup reopen in my case)
- CommandItem : use name/label as value={`${item.name}`}
- CommandItem: onSelect filter list by the element ID then set new value `${currentValue?.id}`

`"use client";
import React, { use, useEffect, useState } from "react";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
} from "@/components/ui/command";
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import { Check, ChevronsUpDown } from "lucide-react";
import { cn } from "@/lib/utils";

export interface DropdownListItem {
  id: number;
  name: string;
}

export interface DropdownProps {
  list: DropdownListItem[];
}

export default function DropdownCommand({ list }: DropdownProps) {
  const [open, setOpen] = useState(false);
  const [value, setValue] = useState("");
  const [search, setSearch] = useState("");

  useEffect(() => {
    if (open) {
      setSearch("");
    }
  }, [open]);

  return (
    <Popover open={open} onOpenChange={setOpen}>
      <PopoverTrigger asChild>
        <Button
          variant="secondary"
          role="combobox"
          aria-expanded={open}
          className="w-[350px] justify-between"
        >
          {value
            ? list.find((item) => `${item.id}` === value)?.name
            : "Choisir une catégorie"}
          <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
        </Button>
      </PopoverTrigger>
      <PopoverContent className="w-[200px] p-0">
        <Command>
          <CommandInput
            placeholder="Search framework..."
            className="h-9"
            value={search}
            onValueChange={setSearch}
          />
          <CommandEmpty>No framework found.</CommandEmpty>
          <CommandGroup>
            {list.map((item) => (
              <CommandItem
                key={`${item.id}`}
                value={`${item.name}`}
                title={item.name}
                onSelect={() => {
                  const currentValue = list.find((el) => el.id === item.id);
                  setValue(
                    `${currentValue?.id}` === value
                      ? ""
                      : `${currentValue?.id}`,
                  );
                  setOpen(false);
                }}
              >
                {item.name}
                <Check
                  className={cn(
                    "ml-auto h-4 w-4",
                    value === `${item.id}` ? "opacity-100" : "opacity-0",
                  )}
                />
              </CommandItem>
            ))}
          </CommandGroup>
        </Command>
      </PopoverContent>
    </Popover>
  );
}
`

**Comment by nstylo** [comment]
+1, also currently bumping into the same issue. I think there should be an option to define the search key.

**Comment by varna** [comment]
What if `onSelect(value: string)` would be changed into `onSelect(value: string, key: string)`?
We could keep using `value` as text for simplicity, but use our own custom identifier onChange to emit proper value outside of component.

I came here to look for solution, cause I tried using:
```js 
value={item.name}
onSelect(value => onChange(
  items.find(item => item.name === value)?.id
))
```
But it seems that value gets slugified or similarly changed before going to onSelect event.

**Comment by McTom234** [comment]
@pacocoursey I found a (in my opinion - works well for me) solution using a id-value map. I made a fork for work and could prepare a small PR within the next hours if you think that a id-value map would be cool, at least as a work-around.

**Comment by jtapeg** [comment]
I noticed that version 0.2.1 still doesn't have keywords

**Comment by snoot-booper** [comment]
Same, I'm trying to use keywords to circumvent the search key problem like suggested here and I see the PR is merged. But 0.2.1 does not include it @pacocoursey

**Comment by kevin-dapps** [comment]
+1

**Comment by afrieirham** [comment]
If anyone looking for a safe workaround, here's mine to consider.

Problem: 
- Attaching `id-name` as value makes the search less accurate especially if the `id` is some unique string id.
- Filtering based on `id-name` is hard since i cant just split based on "-" (my id can have that). 
   - I consider making the `id` length to a fixed number but it's still feel weird setting a random fixed length for `id`
   
Solution: 

- Use stringified JSON as value.
```ts
// use this as Command.Item value
const stringValue = JSON.stringify({
  id: item.id,
  name: item.name,
}).replaceAll('"', "'");

// replaceAll is needed because it won't work with double quotes

```

```tsx
// for the filter function
<Command
  filter={(value, search) => {
    const { name } = JSON.parse(value.replaceAll("'", '"')) as { id: string; name: string; };
    if (name?.includes(search)) return 1;
    return 0;
  }}
></Command>;

```

Note that this is just a temporary workaround. As soon as the `keywords` prop available, I'll switch to it.

You can refer my full implementation [here](https://github.com/afrieirham/carikabel/commit/e101fb4c8a766c8182a09faf51916428748d86bd)

**Comment by divmgl** [comment]
I just now ran into a similar situation too and my mind immediately went to "how do I set the search value?" But after working with this for a bit I think I prefer `cmdk` does this: just set a custom filter. This lets you, the implementer, decide how you want it to work when the defaults don't do what you want.

Here's a contrived example.

```tsx
function CarsDropdown() {
  // Assuming `cars` comes from somewhere above

  // Create a map to avoid O(n) lookups
  const carsMap = cars.reduce( 
    (acc, car) => {
      acc[car.id] = car // Use the unique identifier as the key
      return acc
    },
    {} as Record<string, Car>
  )

  return (
    <Command
      filter={(value, search) => {
        const car = carsMap[value]
        // Basic search but you can extend this
        if (car && car.name.toLowerCase().includes(search.toLowerCase())) return 1
        return 0
      }}
    >
      <CommandGroup>
        {cars.map((car) => (
          <CommandItem
            key={car.id}
            value={car.id}
          >
            {car.name}
          </CommandItem>
        )}
      </CommandGroup>
    </Command>
  )

}
```

**Comment by asontha** [comment]
Any update on this?

**Comment by pacocoursey** [comment]
`keywords` prop is available in `v1.0.0`.

https://github.com/pacocoursey/cmdk/releases/tag/v1.0.0

**Comment by jmathewburns** [comment]
In case this can still help someone: 

My specific issue was that my `id` values were unique, but values were not, causing weirdness with multiple items appearing to be selected. I could deduplicate `value` using the `id`, but I didn't want to make them searchable if they're not shown, as that's it's own kind of weirdness.

 Thanks to this PR: https://github.com/pacocoursey/cmdk/pull/229, we can now use the default `cmdk` filter function, so this was the simplest solution I came up with:

```tsx
<Command filter={(value, search, keywords) => defaultFilter(value.substring(value.indexOf('#') + 1), search, keywords ?? [])}>
    <CommandInput placeholder="Search ..." />
    <CommandList>
        <CommandEmpty>No options found.</CommandEmpty>
        <CommandGroup>
            {options.map((option) => (
                <CommandItem
                    key={option.id}
                    value={option.id + '#' + option.name}

                    onSelect={() => {
                        onSelect(option)
                    }}
                >
                    {option.name}
                </CommandItem>
            ))}
        </CommandGroup>
    </CommandList>
</Command>
```

Simply deduplicate using the id, then pre-process it away before filtering. No custom filtering logic or extra state required
